File: backend/app/schemas/forum.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

ForumCategory = Literal["报到", "生活", "学习", "社团", "其他"]
ForumSort = Literal["latest", "hot", "open"]


class ForumPostCreate(BaseModel):
    title: str = Field(..., min_length=2, max_length=120)
    content: str = Field(..., min_length=5, max_length=2000)
    category: ForumCategory = "其他"


class ForumPostUpdate(BaseModel):
    title: str | None = Field(None, min_length=2, max_length=120)
    content: str | None = Field(None, min_length=5, max_length=2000)
    category: ForumCategory | None = None


class ForumAnswerCreate(BaseModel):
    content: str = Field(..., min_length=1, max_length=2000)


class ForumAnswerUpdate(BaseModel):
    content: str = Field(..., min_length=1, max_length=2000)


class ForumAuthorBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    name: str
    class_name: str


class ForumPostBrief(BaseModel):
    id: UUID
    title: str
    content_preview: str
    category: str
    author: ForumAuthorBrief
    answer_count: int
    has_accepted: bool
    is_closed: bool
    is_pinned: bool
    like_count: int = 0
    liked_by_me: bool = False
    created_at: datetime
    is_mine: bool = False


class ForumAnswerItem(BaseModel):
    id: UUID
    content: str
    author: ForumAuthorBrief
    is_accepted: bool
    like_count: int = 0
    liked_by_me: bool = False
    created_at: datetime
    is_mine: bool = False


class ForumPostDetail(BaseModel):
    id: UUID
    title: str
    content: str
    category: str
    author: ForumAuthorBrief
    author_id: int
    answer_count: int
    has_accepted: bool
    is_closed: bool
    is_pinned: bool
    like_count: int = 0
    liked_by_me: bool = False
    created_at: datetime
    is_mine: bool = False
    answers: list[ForumAnswerItem]

File: frontend/src/types/forum.ts
export type ForumCategory = '报到' | '生活' | '学习' | '社团' | '其他'

export interface ForumAuthor {
  name: string
  class_name: string
}

export interface ForumPostBrief {
  id: string
  title: string
  content_preview: string
  category: ForumCategory
  author: ForumAuthor
  answer_count: number
  has_accepted: boolean
  is_closed: boolean
  is_pinned: boolean
  like_count: number
  liked_by_me: boolean
  created_at: string
  is_mine: boolean
}

export interface ForumAnswer {
  id: string
  content: string
  author: ForumAuthor
  is_accepted: boolean
  like_count: number
  liked_by_me: boolean
  created_at: string
  is_mine: boolean
}

export interface ForumPostDetail {
  id: string
  title: string
  content: string
  category: ForumCategory
  author: ForumAuthor
  author_id: number | null
  answer_count: number
  has_accepted: boolean
  is_closed: boolean
  is_pinned: boolean
  like_count: number
  liked_by_me: boolean
  created_at: string
  is_mine: boolean
  answers: ForumAnswer[]
}

export const FORUM_CATEGORIES: ForumCategory[] = ['报到', '学习', '生活', '社团', '其他']

export const FORUM_CATEGORY_COLORS: Record<ForumCategory, string> = {
  报到: '#b5343a',
  生活: '#4a8c5c',
  学习: '#6b5b95',
  社团: '#c9a96e',
  其他: '#8b7b65',
}
